File: src/exonware/xwsystem/config/performance.py
#exonware/xwsystem/src/exonware/xwsystem/config/performance.py
"""Performance optimization configuration for xwsystem I/O operations.

This module provides configuration options to enable/disable performance
optimizations (parallel index building, append-only logs) with automatic
fallback to original implementations.
"""

from __future__ import annotations

from dataclasses import dataclass
import os


@dataclass
class PerformanceConfig:
    """Configuration for performance optimizations."""
    
    # Parallel index building
    enable_parallel_index: bool = True  # Auto-enabled for large files
    parallel_index_workers: int | None = None  # None = auto (CPU count)
    parallel_index_chunk_size_mb: int = 100  # 100MB chunks
    parallel_index_threshold_mb: int = 200  # Use parallel for files >200MB
    
    # Append-only log for atomic updates
    enable_append_log: bool = True  # Auto-enabled for large files
    append_log_threshold_mb: int = 100  # Use append-only log for files >100MB
    append_log_compaction_threshold_mb: int = 100  # Compact when log > 100MB
    
    # Fallback behavior
    fallback_on_error: bool = True  # Fall back to original if optimization fails
    
    @classmethod
    def from_env(cls) -> PerformanceConfig:
        """Create config from environment variables."""
        return cls(
            enable_parallel_index=os.getenv("XWSYSTEM_PARALLEL_INDEX", "true").lower() == "true",
            parallel_index_workers=int(os.getenv("XWSYSTEM_PARALLEL_WORKERS", "0")) or None,
            parallel_index_chunk_size_mb=int(os.getenv("XWSYSTEM_CHUNK_SIZE_MB", "100")),
            parallel_index_threshold_mb=int(os.getenv("XWSYSTEM_PARALLEL_THRESHOLD_MB", "200")),
            enable_append_log=os.getenv("XWSYSTEM_APPEND_LOG", "true").lower() == "true",
            append_log_threshold_mb=int(os.getenv("XWSYSTEM_APPEND_LOG_THRESHOLD_MB", "100")),
            append_log_compaction_threshold_mb=int(os.getenv("XWSYSTEM_LOG_THRESHOLD_MB", "100")),
            fallback_on_error=os.getenv("XWSYSTEM_FALLBACK", "true").lower() == "true",
        )
    
    @classmethod
    def conservative(cls) -> PerformanceConfig:
        """Conservative config (disable optimizations, use originals)."""
        return cls(
            enable_parallel_index=False,
            enable_append_log=False,
            fallback_on_error=True,
        )
    
    @classmethod
    def aggressive(cls) -> PerformanceConfig:
        """Aggressive config (enable all optimizations, no fallback)."""
        return cls(
            enable_parallel_index=True,
            enable_append_log=True,
            fallback_on_error=False,
        )


# Global config instance
_config: PerformanceConfig | None = None


def get_performance_config() -> PerformanceConfig:
    """Get global performance config."""
    global _config
    if _config is None:
        _config = PerformanceConfig.from_env()
    return _config


def set_performance_config(config: PerformanceConfig) -> None:
    """Set global performance config."""
    global _config
    _config = config


__all__ = [
    "PerformanceConfig",
    "get_performance_config",
    "set_performance_config",
]

File: rust/src/config/performance.rs
// #exonware/xwsystem/rust/src/config/performance.rs
//! Performance optimization configuration for xwsystem I/O operations.
//! 
//! This module provides configuration options to enable/disable performance
//! optimizations (parallel index building, append-only logs) with automatic
//! fallback to original implementations.


use std::collections::HashMap;
use std::sync::{Arc, RwLock};

/// Configuration for performance optimizations.
#[derive(Debug, Clone)]
pub struct PerformanceConfig {
    pub enable_parallel_index: bool,
    pub parallel_index_workers: Option<i64>,
    pub parallel_index_chunk_size_mb: i64,
    pub parallel_index_threshold_mb: i64,
    pub enable_append_log: bool,
    pub append_log_threshold_mb: i64,
    pub append_log_compaction_threshold_mb: i64,
    pub fallback_on_error: bool,
}

impl PerformanceConfig {
    /// Create config from environment variables.
    pub fn from_env() -> Self {
        Self {
            enable_parallel_index: std::env::var("XWSYSTEM_PARALLEL_INDEX")
                .unwrap_or_else(|_| "true".to_string())
                .to_lowercase() == "true",
            parallel_index_workers: std::env::var("XWSYSTEM_PARALLEL_WORKERS")
                .ok()
                .and_then(|s| s.parse().ok())
                .filter(|&v: &i64| v > 0),
            parallel_index_chunk_size_mb: std::env::var("XWSYSTEM_CHUNK_SIZE_MB")
                .unwrap_or_else(|_| "100".to_string())
                .parse()
                .unwrap_or(100),
            parallel_index_threshold_mb: std::env::var("XWSYSTEM_PARALLEL_THRESHOLD_MB")
                .unwrap_or_else(|_| "200".to_string())
                .parse()
                .unwrap_or(200),
            enable_append_log: std::env::var("XWSYSTEM_APPEND_LOG")
                .unwrap_or_else(|_| "true".to_string())
                .to_lowercase() == "true",
            append_log_threshold_mb: std::env::var("XWSYSTEM_APPEND_LOG_THRESHOLD_MB")
                .unwrap_or_else(|_| "100".to_string())
                .parse()
                .unwrap_or(100),
            append_log_compaction_threshold_mb: std::env::var("XWSYSTEM_LOG_THRESHOLD_MB")
                .unwrap_or_else(|_| "100".to_string())
                .parse()
                .unwrap_or(100),
            fallback_on_error: std::env::var("XWSYSTEM_FALLBACK")
                .unwrap_or_else(|_| "true".to_string())
                .to_lowercase() == "true",
        }
    }

    /// Conservative config (disable optimizations, use originals).
    pub fn conservative() -> Self {
        Self {
            enable_parallel_index: false,
            parallel_index_workers: None,
            parallel_index_chunk_size_mb: 100,
            parallel_index_threshold_mb: 200,
            enable_append_log: false,
            append_log_threshold_mb: 100,
            append_log_compaction_threshold_mb: 100,
            fallback_on_error: true,
        }
    }

    /// Aggressive config (enable all optimizations, no fallback).
    pub fn aggressive() -> Self {
        Self {
            enable_parallel_index: true,
            parallel_index_workers: None,
            parallel_index_chunk_size_mb: 100,
            parallel_index_threshold_mb: 200,
            enable_append_log: true,
            append_log_threshold_mb: 100,
            append_log_compaction_threshold_mb: 100,
            fallback_on_error: false,
        }
    }
}

/// Placeholder for backward compatibility.
#[derive(Debug, Clone, Default)]
pub struct PerformanceLimits {}

/// Placeholder for backward compatibility.
#[derive(Debug, Clone, Default)]
pub struct SerializationLimits {}

/// Placeholder for backward compatibility.
#[derive(Debug, Clone, Default)]
pub struct NetworkLimits {}

/// Placeholder for backward compatibility.
#[derive(Debug, Clone, Default)]
pub struct SecurityLimits {}

// Global config instance
use std::sync::OnceLock;

static GLOBAL_CONFIG: OnceLock<Arc<RwLock<Option<PerformanceConfig>>>> = OnceLock::new();

fn get_global_config() -> Arc<RwLock<Option<PerformanceConfig>>> {
    GLOBAL_CONFIG.get_or_init(|| Arc::new(RwLock::new(None))).clone()
}

/// Get global performance config.
pub fn get_performance_config() -> PerformanceConfig {
    let config = get_global_config().read().unwrap();
    config.clone().unwrap_or_else(PerformanceConfig::from_env)
}

/// Set global performance config.
pub fn set_performance_config(config: PerformanceConfig) {
    let mut global = get_global_config().write().unwrap();
    *global = Some(config);
}

/// Placeholder for backward compatibility.
pub fn configure_performance(_kwargs: HashMap<String, String>) {
    // Configuration would be applied here
}

/// Placeholder for backward compatibility.
pub fn get_serialization_limits() -> SerializationLimits {
    SerializationLimits::default()
}

/// Placeholder for backward compatibility.
pub fn get_network_limits() -> NetworkLimits {
    NetworkLimits::default()
}

/// Placeholder for backward compatibility.
pub fn get_security_limits() -> SecurityLimits {
    SecurityLimits::default()
}


